qiskit/utils/units.py

```diff
@@ -25,16 +32,120 @@ def apply_prefix(value: float, unit: str) -> float:
+
+    try:
+        value = float(value)
+    except TypeError as ex:
+        if isinstance(value, ParameterExpression) and value.parameters:
+            return pow(10, prefactors[unit[0]]) * value
+        raise TypeError(f"Input value {value} is not a number.") from ex
+
+    # to avoid round-off error of prefactor
+    _value = Decimal.from_float(value).scaleb(prefactors[unit[0]])
+
+    return float(_value)
+
+
+def detach_prefix(value: float, decimal: Optional[int] = None) -> Tuple[float, str]:
+    """
+    Given a SI unit value, find the most suitable prefix to scale the value.
+
+    For example, the ``value = 1.3e8`` will be converted into a tuple of ``(130.0, "M")``,
+    which represents a scaled value and auxiliary unit that may be used to display the value.
+    In above example, that value might be displayed as ``130 MHz`` (unit is arbitrary here).
+
+    Example:
+
+        >>> value, prefix = detach_prefix(1e4)
+        >>> print(f"{value} {prefix}Hz")
+        10 kHz
+
+    Args:
+        value: The number to find prefix.
+        decimal: Optional. An arbitrary integer number to represent a precision of the value.
+            If specified, it tries to round the mantissa and adjust the prefix to rounded value.
+            For example, 999_999.91 will become 999.9999 k with ``decimal=4``,
+            while 1.0 M with ``decimal=3`` or less.
+
+    Returns:
+        A tuple of scaled value and prefix.
+
+    Note:
+        This may induce tiny value error due to internal representation of float object.
+        See https://docs.python.org/3/tutorial/floatingpoint.html for details.
+
+    Raises:
+        Exception: If the ``value`` is out of range.
+        ValueError: If the ``value`` is not real number.
+        TypeError: If the ``value`` is not a number.
+    """
+    prefactors = {
+        -15: "f",
+        -12: "p",
+        -9: "n",
+        -6: "µ",
+        -3: "m",
+        0: "",
+        3: "k",
+        6: "M",
+        9: "G",
+        12: "T",
+        15: "P",
+    }
+
+    try:
+        value = float(value)
+    except TypeError as ex:
+        raise TypeError(f"Input value {value} is not a number.") from ex
+
+    if not np.isreal(value):
+        raise ValueError(f"Input should be real number. Cannot convert {value}.")
+
+    if np.abs(value) != 0:
+        pow10 = int(np.floor(np.log10(np.abs(value)) / 3) * 3)
     else:
-        raise Exception(f"Could not understand units: {unit}")
+        pow10 = 0
+
+    # to avoid round-off error of prefactor
+    mant = float(Decimal.from_float(value).scaleb(-pow10))
+
+    if decimal is not None:
+        # Corner case handling
+        # For example, 999_999.99 can be rounded to 1000.0 k rather than 1.0 M.
+        mant = np.round(mant, decimal)
+        if mant >= 1000:
+            mant /= 1000
+            pow10 += 3
+
+    if pow10 not in prefactors:
+        raise Exception(f"Value is out of range: {value}")
```

Comment:
SImilar to other function - we should return a proper error (`ValueError`?) rather than `Exception`.

Reply:
Updated in https://github.com/Qiskit/qiskit-terra/pull/6885/commits/1ba3e95770076f2afc7c3c88a4cf43cf8c740e16